### PR TITLE
fix: Fix the incorrect null joined rows for SMJ outer join with join filter

### LINKED ISSUE #10882: SMJ: incorrect result for filtered right outer join
### Describe the bug

It seems DF not correctly handling the right filtered join.

### To Reproduce

```
#[tokio::test]
    async fn test_smj_right_filtered() -> Result<()> {
        let ctx: SessionContext = SessionContext::new();

        let sql = "set datafusion.optimizer.prefer_hash_join = false;";
        let _ = ctx.sql(sql).await?.collect().await?;

        let sql = "set datafusion.execution.batch_size = 100";
        let _ = ctx.sql(sql).await?.collect().await?;

        let sql = "
            select * from (
            with t as (
                select id, id % 5 id1 from (select unnest(range(0,10)) id)
            ), t1 as (
                select id % 10 id, id + 2 id1 from (select unnest(range(0,10)) id)
            )
            select * from  t right join  t1 on t.id1 = t1.id and t.id > t1.id1
            ) order by 1, 2, 3, 4
        ";

        let actual = ctx.sql(sql).await?.collect().await?;

        let expected: Vec<&str> = vec![
            "+----+-----+----+-----+",
            "| id | id1 | id | id1 |",
            "+----+-----+----+-----+",
            "| 5  | 0   | 0  | 2   |",
            "| 6  | 1   | 1  | 3   |",
            "| 7  | 2   | 2  | 4   |",
            "| 8  | 3   | 3  | 5   |",
            "| 9  | 4   | 4  | 6   |",
            "|    |     | 5  | 7   |",
            "|    |     | 6  | 8   |",
            "|    |     | 7  | 9   |",
            "|    |     | 8  | 10  |",
            "|    |     | 9  | 11  |",
            "+----+-----+----+-----+",
        ];
        datafusion_common::assert_batches_eq!(expected, &actual);

        Ok(())
    }
```

### Expected behavior

Test should pass

### Additional context

Found in https://github.com/apache/datafusion-comet/pull/553

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #10882.

### CODE REVIEW

File: datafusion/physical-plan/src/joins/sort_merge_join.rs [hunk 1]
```diff
@@ -860,6 +867,8 @@ impl SMJStream {
                                 self.reservation.shrink(buffered_batch.size_estimation);
                             }
                         } else {
+                            // If the head batch is not fully processed, break the loop.
+                            // Streamed batch will be joined with the head batch in the next step.
```
Comment (viirya) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:871]: Not related to this issue. But as I re-read these code, I add some comments to help us understand it again.

**Comment by alamb** [comment]
Possibly related to https://github.com/apache/datafusion/issues/10881

**Comment by viirya** [comment]
> Possibly related to #10881

Hmm, I don't see smj is explicitly enabled there?

File: datafusion/core/tests/sql/joins.rs [hunk 2]
```diff
@@ -235,3 +235,47 @@ async fn join_change_in_planner_without_sort_not_allowed() -> Result<()> {
...
 }
+
+#[tokio::test]
+async fn test_smj_right_filtered() -> Result<()> {
+    let ctx: SessionContext = SessionContext::new();
+
+    let sql = "set datafusion.optimizer.prefer_hash_join = false;";
+    let _ = ctx.sql(sql).await?.collect().await?;
+
+    let sql = "set datafusion.execution.batch_size = 100";
```
Comment (comphead) [inline datafusion/core/tests/sql/joins.rs:246]: we can move it to slt

File: datafusion/sqllogictest/test_files/sort_merge_join.slt [hunk 3]
```diff
@@ -84,7 +84,6 @@ SELECT * FROM t1 LEFT JOIN t2 ON t1.a = t2.a AND t2.b * 50 <= t1.b
 Alice 100 Alice 1
 Alice 100 Alice 2
 Alice 50 Alice 1
-Alice 50 NULL NULL
```
Comment (comphead) [inline datafusion/sqllogictest/test_files/sort_merge_join.slt:87]: does it mean tests were not correct?

Comment (viirya) [reply datafusion/sqllogictest/test_files/sort_merge_join.slt:87]: Yes, the results were incorrect.

Comment (viirya) [reply datafusion/sqllogictest/test_files/sort_merge_join.slt:87]: For example, as `Alice 50` is joined with `Alice 1`, it should not be joined with nulls.

Comment (comphead) [reply datafusion/sqllogictest/test_files/sort_merge_join.slt:87]: Agree, just checked in DuckDB
```
D CREATE TABLE t1(a text, b int); insert into t1  VALUES ('Alice', 50), ('Alice', 100), ('Bob', 1);
D CREATE TABLE t2(a text, b int);
D insert into t2 VALUES ('Alice', 2), ('Alice', 1);
D SELECT * FROM t1 LEFT JOIN t2 ON t1.a = t2.a AND t2.b * 50 <= t1.b
>
> ;
┌─────────┬───────┬─────────┬───────┐
│    a    │   b   │    a    │   b   │
│ varchar │ int32 │ varchar │ int32 │
├─────────┼───────┼─────────┼───────┤
│ Alice   │    50 │ Alice   │     1 │
│ Alice   │   100 │ Alice   │     1 │
│ Alice   │   100 │ Alice   │     2 │
│ Bob     │     1 │         │       │
└─────────┴───────┴─────────┴───────┘
```

File: datafusion/core/tests/sql/joins.rs (hunk 2 above)
Comment (viirya) [reply datafusion/core/tests/sql/joins.rs]: okay

File: datafusion/physical-plan/src/joins/sort_merge_join.rs [hunk 4]
```diff
@@ -1445,9 +1504,13 @@ fn get_buffered_columns(
 /// `streamed_indices` have the same length as `mask`
 /// `matched_indices` array of streaming indices that already has a join filter match
 /// `scanning_buffered_offset` current buffered offset across batches
+///
```
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:1507]: 👍

File: datafusion/physical-plan/src/joins/sort_merge_join.rs [hunk 5]
```diff
@@ -1313,51 +1354,37 @@ impl SMJStream {
...
                         let null_joined_streamed_batch =
                             RecordBatch::try_new(self.schema.clone(), columns.clone())?;
                         self.output_record_batches.push(null_joined_streamed_batch);
 
-                        // For full join, we also need to output the null joined rows from the buffered side
+                        // For full join, we also need to output the null joined rows from the buffered side.
+                        // Usually this is done by `freeze_buffered`. However, if a buffered row is joined with
+                        // streamed side, it won't be outputted by `freeze_buffered`.
+                        // We need to check if a buffered row is joined with streamed side and output.
+                        // If it is joined with streamed side, but finally fails on the join filter,
```
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:1366]:
```suggestion
                        // If it is joined with streamed side, but doesn't match the join filter,
```

File: datafusion/physical-plan/src/joins/sort_merge_join.rs [hunk 6]
```diff
@@ -1098,49 +1107,52 @@ impl SMJStream {
...
 
     // Produces and stages record batch from buffered indices with corresponding
     // NULLs on streamed side.
     //
     // Applicable only in case of Full join.
-    fn freeze_buffered(&mut self, batch_count: usize) -> Result<()> {
+    fn freeze_buffered(
+        &mut self,
+        batch_count: usize,
+        output_join_filter_fail_batch: bool,
```
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:1122]: I'd say we can invent better naming here, but nothing comes to my mind. Perhaps
`output_joined_not_matched`

And would be nice to have a comments on this boolean flag what it does

File: datafusion/physical-plan/src/joins/sort_merge_join.rs [hunk 7]
```diff
@@ -1098,49 +1107,52 @@ impl SMJStream {
...
-                .iter()
-                .map(|f| new_null_array(f.data_type(), buffered_indices.len()))
-                .collect::<Vec<_>>();
-
-            streamed_columns.extend(buffered_columns);
-            let columns = streamed_columns;
-
-            self.output_record_batches
-                .push(RecordBatch::try_new(self.schema.clone(), columns)?);
+            // For buffered rows which are joined with streamed side but failed on join filter
```
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:1141]:
```suggestion
            // For buffered rows which are joined with streamed side but doesn't satisfy the join filter
```

File: datafusion/physical-plan/src/joins/sort_merge_join.rs [hunk 8]
```diff
@@ -1055,7 +1064,7 @@ impl SMJStream {
                         Some(scanning_idx),
                     );
                 } else {
-                    // Join nulls and buffered row
+                    // Join nulls and buffered row for full join
```
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:1067]:
```suggestion
                    // Join nulls and buffered row for FULL join
```

File: datafusion/physical-plan/src/joins/sort_merge_join.rs [hunk 9]
```diff
@@ -1313,51 +1354,37 @@ impl SMJStream {
...
-                                streamed_columns,
-                            )?;
-                            self.output_record_batches.push(null_joined_buffered_batch);
+                            for i in 0..pre_mask.len() {
+                                let buffered_batch = &mut self.buffered_data.batches
+                                    [chunk.buffered_batch_idx.unwrap()];
+                                let buffered_index = buffered_indices.value(i);
+
+                                if !pre_mask.value(i) {
+                                    // For a buffered row that is joined with streamed side but failed on the join filter,
```
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:1375]:
```suggestion
                                    // For a buffered row that is joined with streamed side but doesn't satisfy the join filter,
```

File: datafusion/physical-plan/src/joins/sort_merge_join.rs [hunk 10]
```diff
@@ -1422,6 +1449,38 @@ fn get_filter_column(
...
+    let mut streamed_columns = streamed_schema
+        .fields()
+        .iter()
+        .map(|f| new_null_array(f.data_type(), buffered_indices.len()))
+        .collect::<Vec<_>>();
+
+    streamed_columns.extend(buffered_columns);
+    let columns = streamed_columns;
+
+    Ok(Some(RecordBatch::try_new(schema.clone(), columns)?))
```
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:1481]: Can we do just 

```
Ok(Some(RecordBatch::try_new(schema.clone(), streamed_columns.extend(buffered_columns))?))
```

File: datafusion/physical-plan/src/joins/sort_merge_join.rs [hunk 11]
```diff
@@ -1098,49 +1107,52 @@ impl SMJStream {
...
-            let columns = streamed_columns;
-
-            self.output_record_batches
-                .push(RecordBatch::try_new(self.schema.clone(), columns)?);
+            // For buffered rows which are joined with streamed side but failed on join filter
+            if output_join_filter_fail_batch {
+                let buffered_indices = UInt64Array::from_iter_values(
+                    buffered_batch.join_filter_failed_idxs.iter().copied(),
+                );
+                if let Some(record_batch) = produce_buffered_null_batch(
```
Comment (comphead) [inline datafusion/physical-plan/src/joins/sort_merge_join.rs:1146]: we may want to have a small helper function to reduce the duplicated call?

**Review by comphead — COMMENTED** [review]
Thanks @viirya 💪

File: datafusion/physical-plan/src/joins/sort_merge_join.rs (hunk 11 above)
Comment (viirya) [reply datafusion/physical-plan/src/joins/sort_merge_join.rs:1146]: I already pull common one to `produce_buffered_null_batch`. You mean all code in the `for` into a function?

Comment (comphead) [reply datafusion/physical-plan/src/joins/sort_merge_join.rs:1146]: I was thinking to have a small helper function to wrap 

```
          if let Some(record_batch) = produce_buffered_null_batch(
                 &self.schema,
                 &self.streamed_schema,
                 &buffered_indices,
                 buffered_batch,
             )? {
                 self.output_record_batches.push(record_batch);
             }
```
 I just realized it is a formatting, the block has 8 lines but in fact its 1 condition and 1 instruction, so lets leave it as is.

File: datafusion/physical-plan/src/joins/sort_merge_join.rs (hunk 10 above)
Comment (viirya) [reply datafusion/physical-plan/src/joins/sort_merge_join.rs:1481]: Okay, it is actually

```
Ok(Some(RecordBatch::try_new(
    schema.clone(),
    streamed_columns,
)?))
```

File: datafusion/physical-plan/src/joins/sort_merge_join.rs (hunk 6 above)
Comment (viirya) [reply datafusion/physical-plan/src/joins/sort_merge_join.rs]: Renamed it.

**Review by comphead — APPROVED** [review]
lgtm thanks @viirya I think its good to go

**Comment by viirya** [comment]
Thank you @alamb @comphead. Let me resolve the conflicts before merging.

**Comment by alamb** [comment]
Thanks @viirya  and @comphead

**Pull request merged**